prov/gni/src/gnix_wait.c

```diff
@@ -256,7 +400,49 @@ static int gnix_wait_control(struct fid *wait, int command, void *arg)
  */
 DIRECT_FN int gnix_wait_wait(struct fid_wait *wait, int timeout)
 {
-	return -FI_ENOSYS;
+	int err = 0, ret;
+	char c;
+	struct gnix_fid_wait *wait_priv;
+
+	GNIX_TRACE(WAIT_SUB, "\n");
+
+	wait_priv = container_of(wait, struct gnix_fid_wait, wait.fid);
+	switch (wait_priv->type) {
+	case FI_WAIT_UNSPEC:
+		gnix_wait_thread_enabled = 1;
+		pthread_cond_signal(&gnix_wait_cond);
+		GNIX_TRACE(WAIT_SUB,
+			   "Calling fi_poll_fd %d timeout %d\n",
+			   wait_priv->fd[WAIT_READ],
+			   timeout);
+		err = fi_poll_fd(wait_priv->fd[WAIT_READ], timeout);
```

Comment:
I based this on the other providers that is the reason behind using fi functions inside the provider.